ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java

```diff
@@ -662,6 +673,55 @@ private void onNonFunctionKey(
             default:
                 if (isWordSeparator(primaryCode)) {
                     handleSeparator(primaryCode);
+                    // retrieves the first two characters before cursor
+                    String previousTwoChars = ic.getTextBeforeCursor(2, 0).toString();
```

Comment:
accessing `ic` is a heavy operation (it basically, needs to communicate with the hosting app, and blocks the main thread).
Try to get that information either from `mLastSpaceTimeStamp`,`TextEntryState` or `mWord`.